### ISSUE TITLE
Error at Deploying Skill Luis app. (New Skill Template) - "Access denied due to invalid subscription key"

### ISSUE THREAD

**Issue by waqarghani** [microsoft/botframework-solutions#1373]
## Project

**Name**: Skills

**Language**: C#

## Description
Error at Deploying Skill Luis app. (New Skill Template)
**The Error i am facing can be seen in the image below as well:** 

luisgen : The term 'luisgen' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Aconext\code_hub\c#\SpeechVAAttempt\SkillAta\SkillAta\SkillAta\SkillAta\Deployment\Scripts\luis_functions.ps1:107 char:2
+     luisgen $luisFile -cs "$($outName)Luis" -o $outFolder
+     ~~~~~~~
+ CategoryInfo          : ObjectNotFound: (luisgen:String) [], CommandNotFoundException
+ FullyQualifiedErrorId : CommandNotFoundException

**I thought may be it is a problem with luisgen updated from 1.2.0 -> 1.3... something and then back to 1.2.0**

**Also the log file shows something like that:**


Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.

LUIS Command Line Interface - ┬® 2018 Microsoft Corporation

Sends a training request for a version of a specified LUIS app. This POST request initiates a request asynchronously. To determine whether the training request is successful, submit a GET request to get training status. Note: The application version is not fully trained unless all the models (intents and entities) are trained successfully or are up to date. To verify training success, get the training status at least once after training is complete.

Usage:
luis train version --appId <appId> --versionId <versionId> [--wait]

Command arguments are:
 --appId <string> (required)    The application ID.                                                                                                                                                                 
 --versionId <string>           The version ID.                                                                                                                                                                     
 (required)                                                                                                                                                                                                         
 --wait <boolean>               Wait for the training to complete                                                                                                                                                   

Configuration and Overrides:
 --appId                        Specifies the public LUIS application id. Overrides the .luisrc value and the LUIS_APP_ID environment variable.                                                                     
 --authoringKey                 Specifies the LUIS authoring  key (from luis.ai portal user settings page). Overrides the .luisrc value and the LUIS_AUTHORING_KEY environment variable.                            
 --subscriptionKey              Specifies the LUIS subscriptionKey. Overrides the .luisrc value and the LUIS_SUBSCRIPTION_KEY environment variable.                                                                 
 --versionId                    Specifies the version id. Overrides the .luisrc value and the LUIS_VERSION_ID environment variable.                                                                                 
 --region                       Specifies the authoring region for all requests. [westus|westeurope|australiaeast] Overrides the .luisrc value and the LUIS_REGION environment variable.                            
 --stdin                        Pull in service keys from stdin in the format of that is the output of: msbot get service                                                                                           
 --prefix                       Appends [luis-apis] prefix to all messages                                                                                                                                          

Global Arguments:
 --help,    -h                  Prints this help file.                                                                                                                                                              
 --version, -v                  Prints the version of this cli tool                                                                                                                                                 
 --force,   -f                  Do not prompt for confirmation, force the operation                                                                                                                                 
 --!                            Dumps all documented commands to the console with descriptions                                                                                                                      

But what has subscription key to do here because i am not even been asked for subscription key in deployment procedure. I am only providing my Authoring key. But i also tried to provide the Azure Portal Resource LUIS generated Key 1 as subscription key as well , but no change.   


## To Reproduce
- Download and install and create a new Skill Template. 
- Deploy your skill using powershell script : .\Deployment\Scripts\deploy.ps1 -appid and -apppassword

## Expected behavior
A running skill bot

## Screenshots
![luisgenerror](https://user-images.githubusercontent.com/29833054/57866629-ef7fb280-77ff-11e9-9b40-f07c7045f540.png)

[deploy_cognitive_models_log.txt](https://github.com/microsoft/botframework-solutions/files/3187816/deploy_cognitive_models_log.txt)

Thank you for your help.

**Comment by waqarghani** [comment]
Sorry...  i found my problem ... it was my mistake ... i was missing in the installation procedure this line:
npm install -g botdispatch ludown@1.2.0 luis-apis qnamaker luisgen@2.0.2 botskills 
I had luisgen but 2.1.0 it was uninstall and 2.0.2 was installed either this was problem or with botskills Absence. I think the first one being more probable as the error is showing luisgen problem. 
Screenshot attached.
Thank you for your support.   
![SolutionLUISGEN](https://user-images.githubusercontent.com/29833054/57913359-bc373500-788c-11e9-979c-822eef8c58ff.png)

**Comment by waqarghani** [comment]
ok i got a little bit hasty there ... pwsh shows everything is done without error ... but the log of cognitive model shows another story attached below ... it is still giving the same error of subscription key .. not getting trained and published so when i run bot it still shows some luis error for work around i went to my eu.luis portal and trained and published my luis manually and the bot is working now. So i am still missing something. 
Log file is exactly the same stating the subscription key error ... and resulting in not publishing or training luis model. 
Thank you once again for your support. Great work! 

[deploy_cognitive_models_log.1.txt](https://github.com/microsoft/botframework-solutions/files/3190623/deploy_cognitive_models_log.1.txt)

**Comment by mdrichardson** [comment]
Possibly related: https://github.com/microsoft/botbuilder-tools/issues/585

Can you verify that you got your authoring key from **eu**.luis.ai and not **www**.luis.ai? Also please verify that your LUIS authoring region is "westeurope".

Possibly related: https://github.com/microsoft/botbuilder-tools/issues/1081#issuecomment-492815757

Can you verify that you're using your authoring key (which is needed for publishing the apps/deploying cognitive models) and not subscription key (which is what you'd use for queries)? You can find it in **eu**.luis.ai > User > Settings:

![image](https://user-images.githubusercontent.com/40401643/57948267-0e308880-7896-11e9-83d1-3d22bbb42959.png)

**Comment by waqarghani** [comment]
Hi @mdrichardson I verfiy you that i always take my authoring key from user/settings from eu.luis.ai page if i want to deploy luis in westeurope , which most of the time i do ... so i am 100 % sure there is no error happening there. And i am pretty sure too my resource group as well as luis location is always europe. Fo example see the Skill with skill template image Resource group from my portal , below: notice how the application insight is in west us but this not be the problem right and then i have also attached also a resource group image from my virtual assistant once again from my portal there i used customize template to change application insight to west europe and only qna services like search and app is in west us because qna doesn't have a deployement location anywhere else then westus. 

![skilleurope](https://user-images.githubusercontent.com/29833054/57966513-2dd9b680-7953-11e9-8805-e4c2ada622de.png)

![VAEurope](https://user-images.githubusercontent.com/29833054/57966516-36ca8800-7953-11e9-9863-29001ee8f9a4.png)

Thank you for your quick response ... after having a conversation with you i can have take two course of actions:

-- Try to customize the template with through this -parametersFile .\Deployment\Resources\parameters.template.json as i did for my virtual assistant and their i change my app insight to west europe too. 

-- Deploy a complete new Virtual Assistant to check if this Luis problem comes there so to confirm may be the problem is independent of Skill template and may be with my procedure.

Finally may be you can suggest something better.

**Comment by mdrichardson** [comment]
@waqarghani I just deployed to West Europe successfully with a brand new VA bot, so there doesn't seem to be anything wrong with the service or commands and I am unable to replicate your error. Can you try again? If it doesn't work, can you try resetting your authoring key? You can do this via User Settings.

Success:
![image](https://user-images.githubusercontent.com/40401643/58033930-46280d80-7ada-11e9-8d66-3c21d8b7442e.png)

![image](https://user-images.githubusercontent.com/40401643/58033951-53dd9300-7ada-11e9-9055-14ac04d67d00.png)

**Comment by mdrichardson** [comment]
@waqarghani Are you still having issues with this?

**Comment by waqarghani** [comment]
@mdrichardson  can you please give me this week, i want to try to deploy this before you close this issue. But i am a little bit busy ? Thank you for your response and patience.

**Comment by mdrichardson** [comment]
@waqarghani No problem.

**Comment by waqarghani** [comment]
Hi @mdrichardson ,

first of all sorry to make you wait for so long... the error still persists : i am going to paste the error below, just for the sake of rechecking:


Access denied due to invalid subscription key. Make sure you are subscribed to an API you are trying to call and provide the right key.

LUIS Command Line Interface - ┬® 2018 Microsoft Corporation

Sends a training request for a version of a specified LUIS app. This POST request initiates a request asynchronously. To determine whether the training request is successful, submit a GET request to get training status. Note: The application version is not fully trained unless all the models (intents and entities) are trained successfully or are up to date. To verify training success, get the training status at least once after training is complete.

Usage:
luis train version --appId <appId> --versionId <versionId> [--wait]

Command arguments are:
 --appId <string> (required)    The application ID.                                                                                                                                                                 
 --versionId <string>           The version ID.                                                                                                                                                                     
 (required)                                                                                                                                                                                                         
 --wait <boolean>               Wait for the training to complete                                                                                                                                                   

Configuration and Overrides:
 --appId                        Specifies the public LUIS application id. Overrides the .luisrc value and the LUIS_APP_ID environment variable.                                                                     
 --authoringKey                 Specifies the LUIS authoring  key (from luis.ai portal user settings page). Overrides the .luisrc value and the LUIS_AUTHORING_KEY environment variable.                            
 --subscriptionKey              Specifies the LUIS subscriptionKey. Overrides the .luisrc value and the LUIS_SUBSCRIPTION_KEY environment variable.                                                                 
 --versionId                    Specifies the version id. Overrides the .luisrc value and the LUIS_VERSION_ID environment variable.                                                                                 
 --region                       Specifies the authoring region for all requests. [westus|westeurope|australiaeast] Overrides the .luisrc value and the LUIS_REGION environment variable.                            
 --stdin                        Pull in service keys from stdin in the format of that is the output of: msbot get service                                                                                           
 --prefix                       Appends [luis-apis] prefix to all messages                                                                                                                                          

Global Arguments:
 --help,    -h                  Prints this help file.                                                                                                                                                              
 --version, -v                  Prints the version of this cli tool                                                                                                                                                 
 --force,   -f                  Do not prompt for confirmation, force the operation                                                                                                                                 
 --!                            Dumps all documented commands to the console with descriptions  

For me this error is in the file (**deploy_cognitive_models_log.txt**) in Deployement folder. It results in LUIS model being not trained and publish so i have to go to my portal and click train and publish as a work around. 

Just to clarify i followed this procedure, to deploy the calander skill:

1. Zip downloaded the git folder of Bot-framework solutions https://github.com/microsoft/botframework-solutions
2. Copied the calander skill and pasted it separately as i want to deploy calander skill only as test. 
3. i have already npm'd the dependencies globally using: npm install -g botdispatch ludown@1.2.0 luis-apis qnamaker luisgen@2.0.2 botskills
4. I go to app Registration in portal and create an app save its id and password.
5. With help of .\Deployment\Scripts\deploy.ps1 -appId blablabla -appPassword "blablabla" i start the deployment to azure portal
     a- I did deployment before reseting my LUIS key and after LUIS key, so both of these steps don't result in removing the error, unfortunately. 
6. pswh.exe doesn't show any error and updates, but then you go to this log and see this error as well as on portal you see your models are not trained and published. 

My question is Log mention nothign about Authoring key but something regarding subscription key ... i don't understand .. still.. well i have given you the status. 

I thank you for your support and effort.

**Comment by mdrichardson** [comment]
@waqarghani No problem. Glad you were able to test. I just tried to repro this again in both westus and westeurope and both of them worked just fine. No errors in the log and luis models are trained and published.

[Here's an explanation on Authoring vs. Subscription(Endpoint) keys](https://docs.microsoft.com/en-us/azure/cognitive-services/luis/luis-concept-keys)

I believe the fix has been found in [this issue](https://github.com/microsoft/botframework-solutions/issues/1430).

Did you clone the Solutions/AI repo in the last 2 days? You can try replacing `Deployment/Scripts/update_cognitive_models.ps1` with [this](https://raw.githubusercontent.com/microsoft/botframework-solutions/master/templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Deployment/Scripts/update_cognitive_models.ps1)

**Comment by mdrichardson** [comment]
@waqarghani Are you still experiencing this issue?

**Comment by waqarghani** [comment]
Hi @mdrichardson , i am really sorry that i am asking again for the time, i know it has been 10 days that i didn't responded to post. But once again before you close it i want to try it. Give me until tomorrow i will read things very carefully, and probably, hopefully close it as well. Cheers thanks for your patience

**Comment by mdrichardson** [comment]
@waqarghani No problem. We usually like to try getting all of the issues closed before a big release, but it's not a big deal. Let's get things working for you!

**Comment by darrenj** [comment]
We have [this](https://github.com/microsoft/botframework-solutions/blob/master/docs/howto/assistant/updatedeploymentscripts.md) documentation item to help guide on updating deployment scripts.

**Comment by mdrichardson** [comment]
@waqarghani 	Are you still experiencing this issue?

**Comment by mdrichardson** [comment]
I am closing this issue as it doesn't seem to be a blocking issue. Feel free to reopen if you have need to get this resolved.
